### PR TITLE
Add typescript support, #537

### PR DESCRIPTION
**Pull request by wayneparrott**
This PR is the initial implementation of #537, adding basic TypeScript support. 
It includes:
* type declaration files: see types/*.d.ts
* dtslint tests, see test/types
* updates to both readme files with new sections for use with TypeScript
* updated package.json with "types" property that points to types/index.d.ts, integrated dtslint to run as part of the test script, and eslint includes the types/ folder.
* tsconfig.json with tsc compiler options
* rosgen_tsd/index.js - script that runs as part of generate_messages.js to update the types/interfaces.d.ts with type aliases

### CODE REVIEW

**Comment by wayneparrott** [comment]
Bummer! Failed during the build phase:
 "Library not loaded: @rpath/librcl.dylib"

I successfully build and install on my Linux system with ros2 eloquent using the same node version, 10.16.1. 

Thoughts on how to move forward?

**Comment by minggangw** [comment]
Hi thanks for your PR, that's something. For the error that happened on CircleCI, I think it not caused by your code, I suggest you could replace the [latest stable packages](https://ci.ros2.org/view/packaging/) of ROS2 with Eloquest packages. The reason why I choose the latest stable packages against `develop` branch is that we want to find the incompatibility with rcl library as early as possible, but the downside is that the package tends to be kind of unstable, e.g. the current failure on CIs.

So I suggest you could change the scripts used by the CIs to use Eloquest directly, please reference the PR https://github.com/RobotWebTools/rclnodejs/pull/538, meanwhile, please don't forget to add branch `typescript` to be built, e.g. [here](https://github.com/RobotWebTools/rclnodejs/blob/develop/.travis.yml#L8), otherwise, the bots will not be triggered. As your code only contains pure JavaScript, I deem that passing the Linux platform is enough to verify the functionality. Thanks!

**Comment by wayneparrott** [comment]
Hitting an issue with the mac os build when running generate_messages.js a dll is not being found. I merged #538 to use eloquent release. Would appreciate any guidance you can provide for resolving this matter.

/Users/distiller/RobotWebTools/rclnodejs/node_modules/bindings/bindings.js:121
        throw e;
        ^

Error: dlopen(/Users/distiller/RobotWebTools/rclnodejs/build/Release/rclnodejs.node, 1): Library not loaded: @rpath/librcl.dylib
  Referenced from: /Users/distiller/RobotWebTools/rclnodejs/build/Release/rclnodejs.node
  Reason: image not found
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:807:18)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at bindings (/Users/distiller/RobotWebTools/rclnodejs/node_modules/bindings/bindings.js:112:48)
    at Object.<anonymous> (/Users/distiller/RobotWebTools/rclnodejs/lib/logging.js:18:38)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! rclnodejs@0.10.4 postinstall: `node scripts/generate_messages.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the rclnodejs@0.10.4 postinstall script.

**Comment by minggangw** [comment]
>  1): Library not loaded: @rpath/librcl.dylib

It seems that the library cannot be found. It's weird that it passed with #538 with the same configuration. I am going to check it, don't worry.

File: rostsd_gen/index.js [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+		}
+	}
+}
+*/
+
+'use strict';
+
+const path = require('path');
+const fs = require('fs');
+const rclnodejs = require('../index.js');
```
Comment (minggangw) [inline rostsd_gen/index.js]: I suspect this line causes the runtime error when loading the dynamic library. I haven't investigated why this only happens on macOS, probably we executed:

``` npm install --python=python2.7```

**Comment by wayneparrott** [comment]
Making some progress now with your previous hint re: loading rclnodejs.js. That was the root cause of the problem with not being able to locate a dll on mac osx. I removed the import of the rclnodejs.js module and success. Thanks for the help. 

While moving forward, still not passing CI tests due to 2 test fails. I will try to look into them later today as they appear to be basic test failures and not related to any of the typescript-based changes.

  1) Cross-language interaction
       Node.js publisher
         Cpp subscription should receive msg from Node.js publisher:
     Error: Timeout of 60000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/Users/distiller/RobotWebTools/rclnodejs/test/test-cross-lang.js)
  

  2) Multiple nodes interation testing
       Publisher/Subscription
         Node.js publisher - Python and Cpp subscription:
     Error: Timeout of 60000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/Users/distiller/RobotWebTools/rclnodejs/test/test-multi-nodes.js)

**Comment by minggangw** [comment]
The tests targeted at interaction with other clients (Python/C++) become flaky sometimes, this happens on the CI machines, not locally. But I still like to retain them because they can verify the compatibility with other clients (some potential changes may occur). I am going to re-run the bots again, hopefully it can pass this time :smile:

**Review by minggangw — COMMENTED** [review]
I think you submitted the binary file `test/cpp/listener` which is for linux caused the error, you may add it mistakenly. Would you please remove it and run again?

**Comment by wayneparrott** [comment]
Inching towards passing CI. :)
Thanks for the tip on the listener exe which I accidentally checked in. 
Now if we can get the Windows platform CI to pass.

**Comment by minggangw** [comment]
Oops, this error happens on `develop` branch constantly as well, which reports:

``` bash
failed to load shared library of rmw implementation. Exception: Cannot load library: c:\ros2-windows\bin/rmw_fastrtps_cpp.dll
```

**Comment by minggangw** [comment]
The problem should be fixed by #544, please rebase your PR and have a another run :smile:

**Comment by wayneparrott** [comment]
**WOOT!** CI finally passed.
Thanks for all of your assistance. Future PRs will go much more smoothly as I think I have the process down.

File: .circleci/config.yml [hunk 2]
```diff
@@ -24,7 +24,7 @@ jobs:
     - run: brew install openssl
     - run: brew install asio console_bridge
     - run: python3 -m pip install catkin_pkg empy git+https://github.com/lark-parser/lark.git@0.7d pyparsing pyyaml setuptools argcomplete colcon-common-extensions numpy
-    - run: mkdir -p ~/ros2_install && cd ~/ros2_install && wget https://ci.ros2.org/view/packaging/job/packaging_osx/lastStableBuild/artifact/ws/ros2-package-osx-x86_64.tar.bz2 && tar xf ros2-package-osx-x86_64.tar.bz2
+    - run: mkdir -p ~/ros2_install && cd ~/ros2_install && wget https://github.com/ros2/ros2/releases/download/release-eloquent-20191122/ros2-eloquent-20191122-macos-amd64.tar.bz2 && tar xf ros2-eloquent-20191122-macos-amd64.tar.bz2
```
Comment (minggangw) [inline .circleci/config.yml:27]: We may change back to the nightly build when we merge back to the `developer` branch.

File: .travis.yml [hunk 3]
```diff
@@ -5,9 +5,7 @@ sudo: required
 
 branches:
   only:
-  - develop
-  - master
-  - actionlib
+  - typescript
```
Comment (minggangw) [inline .travis.yml:8]: We may add `developer` branch later

File: package.json [hunk 4]
```diff
@@ -1,20 +1,22 @@
 {
   "name": "rclnodejs",
-  "version": "0.10.3",
+  "version": "0.10.4",
```
Comment (minggangw) [inline package.json]: I think we can bump to 0.11.0, as we offer a new feature from this version.

File: rostsd_gen/index.js [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+// scan generated files, collect pkg and msg info
+function getPkgInfos(generatedRoot) {
+  let pkgInfos = [];
+  const rootDir = generatedRoot;
+  let pkgs = fs.readdirSync(rootDir);
+
+  for (let pkg of pkgs) {
+
+    if (pkg.endsWith('.json')) continue;
+    //if (pkg.startsWith('actionlib')) continue;
```
Comment (minggangw) [inline rostsd_gen/index.js]: Could this line be deleted?

File: rostsd_gen/index.js [hunk 6]
```diff
@@ -0,0 +1,505 @@
...
+
+    const pkgPath = path.join(rootDir, pkg);
+    const files = fs.readdirSync(pkgPath);
+
+    for (let filename of files) {
+      const typeClass = fileName2Typeclass(filename);
+
+      if (typeClass.type === 'srv') { // skip __srv__<action>
+        if (!typeClass.name.endsWith('Request') && !typeClass.name.endsWith('Response')) {
+          //console.log('skip:', typeClass.name);
```
Comment (minggangw) [inline rostsd_gen/index.js]: Ditto

File: rostsd_gen/index.js [hunk 7]
```diff
@@ -0,0 +1,505 @@
...
+`;
+    fs.writeSync(fd, pkgNamespaceTemplate);
+
+    let curNS = null;
+    for (const msgInfo of pkgInfo.messages) {
+
+      // omit namespaces that are not 'msg' or 'srv'
+      // e.g., action_tutorials.action.FibonacciActionXXX
+      //if (msgInfo.typeClass.type.startsWith('action')) continue;
+      // if (msgInfo.typeClass.name.includes('Fibonacci')) continue;
```
Comment (minggangw) [inline rostsd_gen/index.js]: Are these 4 lines still meaningful?

File: rostsd_gen/index.js [hunk 8]
```diff
@@ -0,0 +1,505 @@
...
+    new typeClass() :
+    undefined;
+}
+
+
+const tsd_generator = {
+  // version() {
+  //   // eslint-disable-next-line
+  //   return fse.readJsonSync(path.join(__dirname, 'generator.json')).version;
+  // },
```
Comment (minggangw) [inline rostsd_gen/index.js]: I think we could delete the comment if we are not going to expose it.

File: rostsd_gen/index.js [hunk 9]
```diff
@@ -0,0 +1,505 @@
...
+  ],
+  "baseType": {
+    "pkgName": "sensor_msgs",
+    "type": "LaserScan",
+    "stringUpperBound": null,
+    "isPrimitiveType": false
+  },
+  "msgName": "LaserScan"
+}
+*/
```
Comment (minggangw) [inline rostsd_gen/index.js]: Could we delete this note?

File: types/interfaces.d.ts [hunk 10]
```diff
@@ -0,0 +1,1796 @@
...
+    visualization_msgs.msg.InteractiveMarkerFeedback |
+    visualization_msgs.msg.InteractiveMarkerInit |
+    visualization_msgs.msg.InteractiveMarkerPose |
+    visualization_msgs.msg.InteractiveMarkerUpdate |
+    visualization_msgs.msg.Marker |
+    visualization_msgs.msg.MarkerArray |
+    visualization_msgs.msg.MenuEntry |
+    visualization_msgs.srv.GetInteractiveMarkers_Request |
+    visualization_msgs.srv.GetInteractiveMarkers_Response;
+}
```
Comment (minggangw) [inline types/interfaces.d.ts]: I have a question here, what's going to happen if a user creates a `.msg`/`.srv` defined by himself? how can offer the type information?

**Review by minggangw — APPROVED** [review]
Thanks for the colossal PR and add this great feature! Some nit comments.

File: .circleci/config.yml (hunk 2 above)
Comment (wayneparrott) [reply .circleci/config.yml:27]: +1

File: .travis.yml (hunk 3 above)
Comment (wayneparrott) [reply .travis.yml:8]: +1

File: package.json (hunk 4 above)
Comment (wayneparrott) [reply package.json]: OK. I'll make this change. to 0.11.0

File: rostsd_gen/index.js (hunk 5 above)
Comment (wayneparrott) [reply rostsd_gen/index.js]: +1, I'll make this change.

File: rostsd_gen/index.js (hunk 6 above)
Comment (wayneparrott) [reply rostsd_gen/index.js]: +1, I'll remove the commented out line.

File: rostsd_gen/index.js (hunk 7 above)
Comment (wayneparrott) [reply rostsd_gen/index.js]: +1, will remove.

File: rostsd_gen/index.js (hunk 8 above)
Comment (wayneparrott) [reply rostsd_gen/index.js]: +1, I'll remove this code.

File: rostsd_gen/index.js (hunk 9 above)
Comment (wayneparrott) [reply rostsd_gen/index.js]: +1, I'll remove the tmp developer notes.

File: types/interfaces.d.ts (hunk 10 above)
Comment (wayneparrott) [reply types/interfaces.d.ts]: My understanding is that when a user creates a custom msg or srv, he must run rclnodejs.regenerateAll() to generate the new js structures. But unfortunately I did not account for regenerateAll() to additionally generate a new interfaces.d.ts updated with new type info for the new msg or srv. My BAD as I only modified the scripts/generate_messages.js#generateAll() to generate the interface.d.ts file which is only run at installation time.

To fix this missing functionality my plan is to update the regenerateAll() function to also generate the interface.d.ts file. Do you agree?

Alternatively, I can add the logic for generating the interface.d.ts file as part of the rosidl_gen/index.js#generateAll() function. 

Thoughts? 

I'll make the changes and test tomorrow.

File: rostsd_gen/index.js (hunk 1 above)
Comment (wayneparrott) [reply rostsd_gen/index.js]: You were correct. I removed line #34 which resolved this problem. Thx!

File: types/interfaces.d.ts (hunk 10 above)
Comment (minggangw) [reply types/interfaces.d.ts]: > To fix this missing functionality my plan is to update the regenerateAll() function to also generate the interface.d.ts file. Do you agree?

I agree we should update the generated interface when any change happens. What's more, as the `types/interfaces.d.ts` is generated dynamically, we'd better not to submit this into repo and it only exists locally. Thanks!

File: rostsd_gen/index.js (hunk 5 above)
Comment (wayneparrott) [reply rostsd_gen/index.js]: fixed.

File: rostsd_gen/index.js (hunk 6 above)
Comment (wayneparrott) [reply rostsd_gen/index.js]: fixed.

File: rostsd_gen/index.js (hunk 7 above)
Comment (wayneparrott) [reply rostsd_gen/index.js]: fixed.

File: rostsd_gen/index.js (hunk 8 above)
Comment (wayneparrott) [reply rostsd_gen/index.js]: fixed.

File: rostsd_gen/index.js (hunk 9 above)
Comment (wayneparrott) [reply rostsd_gen/index.js]: fixed.

File: types/interfaces.d.ts (hunk 10 above)
Comment (wayneparrott) [reply types/interfaces.d.ts]: fixed. See index.js#regenerateAll() which calls the tsd generator after generating the javascript code.

**Pull request merged**